### PR TITLE
Extending Workbox blog post

### LINKED ISSUE #3964: content: Extending Workbox: Custom Strategies & Plugins [2020-10-22]
* [thread](https://groups.google.com/a/google.com/g/webdev-content/c/Dq06nXsnVWo)
* [draft](https://docs.google.com/presentation/d/1fnp_nAuMkuP_h2QjtxJ7Lg8TT-DHT3V7l9oQzHiyOMc/edit?usp=sharing)

### PR DESCRIPTION
**Pull request by jeffposnick**
Fixes #3964

This content was previously reviewed by @kaycebasques.

Changes proposed in this pull request:

- Blog post to accompany my CDS talk. 
- Currently marked as `draft: true` ~~but needs to go live prior to my talk premiere.~~ (in terms of timing, it's okay if the URL works but the article isn't "public")

### CODE REVIEW

File: src/site/content/en/blog/extending-workbox/index.md [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+subhead: >
+  Write your own reusable, shareable strategies and plugins.
+description: >
+  Write your own reusable, shareable strategies and plugins.
+date: 2020-12-10
+thumbnail: thumbnail.png
+hero: hero.png
+alt: The Workbox logo.
+draft: true
+tags:
```
Comment (jeffposnick) [inline src/site/content/en/blog/extending-workbox/index.md:12]: Are there any other tags to use for the CDS-related blog posts?

File: src/site/content/en/blog/extending-workbox/index.md [hunk 2]
```diff
@@ -0,0 +1,379 @@
+---
+title: Extending Workbox
+subhead: >
+  Write your own reusable, shareable strategies and plugins.
+description: >
+  Write your own reusable, shareable strategies and plugins.
+date: 2020-12-10
```
Comment (kaycebasques) [inline src/site/content/en/blog/extending-workbox/index.md:7]:
```suggestion
  Write your own reusable, shareable strategies and plugins.
authors: jeffposnick
date: 2020-12-10
```
Let's attribute this post to you, yes?

File: src/site/content/en/blog/extending-workbox/index.md (hunk 1 above)
Comment (kaycebasques) [reply src/site/content/en/blog/extending-workbox/index.md:12]: There's no CDS-specific posts but I'll check if there are other general tags that are relevant here: https://github.com/GoogleChrome/web.dev/blob/master/src/site/_data/tagsData.json

File: src/site/content/en/blog/extending-workbox/index.md [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+thumbnail: thumbnail.png
+hero: hero.png
+alt: The Workbox logo.
+draft: true
+tags:
+  - blog
+  - service-workers
+---
```
Comment (kaycebasques) [inline src/site/content/en/blog/extending-workbox/index.md:15]:
```suggestion
  - service-worker
  - offline
  - network
  - progressive-web-apps
  - performance
---
```

File: src/site/content/en/blog/extending-workbox/index.md [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+But what if you wanted to go beyond those common caching scenarios? Let's cover
+writing your own custom caching strategies.
+[Workbox v6](https://github.com/GoogleChrome/workbox/releases/tag/v6.0.0) offers
+a new `Strategy` base class that sits in front of lower-level APIs, like
+[Fetch](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API) and
+[Cache Storage](https://developer.mozilla.org/en-US/docs/Web/API/CacheStorage).
+You can extend the `Strategy` base class, and then implement your own logic in
+the `_handle()` method.
+
+### DedupeNetworkFirst
```
Comment (kaycebasques) [inline src/site/content/en/blog/extending-workbox/index.md:55]:
```suggestion
### Handle simultaneous, duplicate requests with DedupeNetworkFirst
```
Didn't catch this during the early review: Let's update these sections to follow the "X with Y" pattern, where X is the general task that's solved and Y is the name of the feature?

File: src/site/content/en/blog/extending-workbox/index.md [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+}
+```
+
+{% Aside %}
+This code assumes that all requests for the same URL can be
+satisfied with the same response, which won't always be the case if cookies or
+session state information comes into play.
+{% endAside %}
+
+### CacheNetworkRace
```
Comment (kaycebasques) [inline src/site/content/en/blog/extending-workbox/index.md:109]:
```suggestion
### Create a race between the cache and network with CacheNetworkRace
```
Not sure how to summarize this section… this suggestion is probably good enough (unless you've got something better)

File: src/site/content/en/blog/extending-workbox/index.md [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+
+Here's another example of a custom strategy—one that's a twist on
+stale-while-revalidate, where both the network and cache are checked at the same
+time, with a race to see which will return a response first.
```
Comment (kaycebasques) [inline src/site/content/en/blog/extending-workbox/index.md:113]: In practice would the network ever be faster here? (just curious; the thought just came to me; no action needed)

File: src/site/content/en/blog/extending-workbox/index.md [hunk 7]
```diff
@@ -0,0 +1,379 @@
...
+            !results[1].value) {
+          reject(results[0].reason);
+        }  
+      });
+    });
+  }
+}
+```
+
+## Using the StategyHandler
```
Comment (kaycebasques) [inline src/site/content/en/blog/extending-workbox/index.md:148]:
```suggestion
## StategyHandler: the recommended approach for creating custom strategies
```
`Using the StrategyHandler` doesn't give me a clue about why I would want to use it

File: src/site/content/en/blog/extending-workbox/index.md [hunk 8]
```diff
@@ -0,0 +1,379 @@
...
+```
+
+A properly written strategy should automatically work with all plugins as well.
+This applies to the standard plugins that Workbox provides, like the one that
+handles
+[cache expiration](https://developers.google.com/web/tools/workbox/modules/workbox-expiration).
+But you're not limited to using the standard set of plugins! Another great way
+to extend Workbox is to write your own reusable plugins.
+
+## Custom Plugins
```
Comment (kaycebasques) [inline src/site/content/en/blog/extending-workbox/index.md:227]:
```suggestion
## Custom plugins
```
Nitpick: sentence case is our site-wide convention

File: src/site/content/en/blog/extending-workbox/index.md [hunk 9]
```diff
@@ -0,0 +1,379 @@
...
+
+Taking a step back, what is a Workbox plugin, and why would you write your own?
+A plugin doesn't fundamentally change the order of network and cache operations
+performed by a strategy. Instead, it allows you to add in extra code that will
+be run at critical points in the lifetime of a request, like when a network
+request fails, or when a cached response is about to be returned to the page.
+
+### Lifecycle Event Overview
```
Comment (kaycebasques) [inline src/site/content/en/blog/extending-workbox/index.md:235]:
```suggestion
### Lifecycle event overview
```

File: src/site/content/en/blog/extending-workbox/index.md [hunk 10]
```diff
@@ -0,0 +1,379 @@
...
+
+When writing your own plugin, you'll only implement callbacks for the limited
+number of events that match your purpose—there's no need to add in callbacks for
+_all_ of the possible events. Additionally, it's up to you whether you implement
+your plugin as an
+<code>[Object](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object)</code>
+with properties that match the lifecycle event names, or as a class that exposes
+methods with those names.
+
+### FallbackOnErrorPlugin
```
Comment (kaycebasques) [inline src/site/content/en/blog/extending-workbox/index.md:310]:
```suggestion
### Lifecycle events example: FallbackOnErrorPlugin
```

**Review by kaycebasques — COMMENTED** [review]
@jeffposnick thank you for the tidy PR. Please also:

- [ ] reference this page from web.dev/reliable (edit `/src/site/_data/paths/reliable.json`)

**Review by kaycebasques — CHANGES_REQUESTED** [review]
whoops, meant to set the last review as "changes requested"

File: src/site/content/en/blog/extending-workbox/index.md (hunk 6 above)
Comment (jeffposnick) [reply src/site/content/en/blog/extending-workbox/index.md:118]: Not usually, but there are some, e.g., old Android devices with really bad flash storage that might be on a fast network.

**Comment by jeffposnick** [comment]
I've edited `/src/site/_data/paths/reliable.json` to add my article as [per your request](https://github.com/GoogleChrome/web.dev/pull/4296#pullrequestreview-538089954), but it's not showing up in the list at https://deploy-preview-4296--web-dev-staging.netlify.app/reliable/

Is that because my article is checked in to the file system under `blog` at `src/site/content/en/blog/extending-workbox/index.md` instead of under `reliable`, e.g. `src/site/content/en/reliable/extending-workbox/index.md`?

**Comment by jeffposnick** [comment]
Hi @kaycebasques—this is still waiting on your re-review/approval.

**Review by kaycebasques — APPROVED** [review]

**Pull request merged**